internal/tools/aws/client.go

```diff
@@ -24,43 +39,72 @@ type AWS interface {
 
 	TagResource(resourceID, key, value string, logger log.FieldLogger) error
 	UntagResource(resourceID, key, value string, logger log.FieldLogger) error
-	IsValidAMI(AMIImage string) (bool, error)
+	IsValidAMI(AMIImage string, logger log.FieldLogger) (bool, error)
+}
+
+// NewAWSClientWithConfig returns a new instance of Client with a custom configuration.
+func NewAWSClientWithConfig(config *aws.Config) *Client {
+	return &Client{
+		config: config,
+		mux:    &sync.Mutex{},
+	}
+}
+
+// Service hold AWS clients for each service.
+type Service struct {
+	acm            acmiface.ACMAPI
+	ec2            ec2iface.EC2API
+	iam            iamiface.IAMAPI
+	rds            rdsiface.RDSAPI
+	s3             s3iface.S3API
+	route53        route53iface.Route53API
+	secretsManager secretsmanageriface.SecretsManagerAPI
+}
+
+// NewService creates a new instance of Service.
+func NewService(sess *session.Session) *Service {
+	return &Service{
+		acm:            acm.New(sess),
+		iam:            iam.New(sess),
+		rds:            rds.New(sess),
+		s3:             s3.New(sess),
+		route53:        route53.New(sess),
+		secretsManager: secretsmanager.New(sess),
+		ec2:            ec2.New(sess),
+	}
 }
 
 // Client is a client for interacting with AWS resources.
 type Client struct {
-	api   api
-	store model.InstallationDatabaseStoreInterface
+	store   model.InstallationDatabaseStoreInterface
+	service *Service
+	config  *aws.Config
+	mux     *sync.Mutex
 }
 
-// api mocks out the AWS API calls for testing.
-type api interface {
-	getRoute53Client() (*route53.Route53, error)
-	changeResourceRecordSets(*route53.Route53, *route53.ChangeResourceRecordSetsInput) (*route53.ChangeResourceRecordSetsOutput, error)
-	listResourceRecordSets(*route53.Route53, *route53.ListResourceRecordSetsInput) (*route53.ListResourceRecordSetsOutput, error)
-	listHostedZones(*route53.Route53, *route53.ListHostedZonesInput) (*route53.ListHostedZonesOutput, error)
-	listTagsForResource(*route53.Route53, *route53.ListTagsForResourceInput) (*route53.ListTagsForResourceOutput, error)
-
-	getEC2Client() (*ec2.EC2, error)
-	tagResource(*ec2.EC2, *ec2.CreateTagsInput) (*ec2.CreateTagsOutput, error)
-	untagResource(*ec2.EC2, *ec2.DeleteTagsInput) (*ec2.DeleteTagsOutput, error)
-	describeImages(svc *ec2.EC2, input *ec2.DescribeImagesInput) (*ec2.DescribeImagesOutput, error)
-
-	getACMClient() (*acm.ACM, error)
-	listCertificates(*acm.ACM, *acm.ListCertificatesInput) (*acm.ListCertificatesOutput, error)
-	listTagsForCertificate(*acm.ACM, *acm.ListTagsForCertificateInput) (*acm.ListTagsForCertificateOutput, error)
-}
+// Service contructs an AWS session if not yet successfully done and returns AWS clients.
+func (c *Client) Service(logger log.FieldLogger) *Service {
+	if c.service == nil {
+		sess, err := NewAWSSessionWithLogger(c.config, logger)
```

Comment:
Service() must be called with `logger` every time, however then `session` is cached in the `Client` object, which means that subsequent calls with a different `logger` will return a `*Service` with the logger initialized the last time, ignoring the one passed in. This would be unexpected behavior. Either the logger should change when it is passed in, or the logger should be part of the Client as outlined in the other comment on this method.

Reply:
> Service() must be called with logger every time, however after the first call session is cached in the Client object, which means that subsequent calls with a different logger will return a *Service with the logger initialized the first time, ignoring the one passed in. This would be unexpected behavior. 

That's not how the above code works. Client may be cached at first call if `NewAWSSessionWithLogger()` does not return an error, otherwise it will return a new service without logger. I don't see how this will produce an unexpected behaviour. 

> Either the logger should change when it is passed in, or the logger should be part of the Client as outlined in the other comment on this method.

I agree in part with that, but I was asked to revert this exactly suggested behaviour https://github.com/mattermost/mattermost-cloud/pull/159#discussion_r389662499

What do you mean by `logger should change when it is passed in`?

Reply:
NewAWSSessionWithLogger is only called if `c.service` is `nil`, otherwise `c.service` is `nil`, which will include the logger from the last time `c.Service()` was called.

Reply:
So if you reuse the client, and call `c.Service(logger)`, `logger` may be the old value passed the first time `c.Service()` is called, instead of the new provided value, which I can't imagine is the expected or intended behavior.

Reply:
So for instance when `c.Service(logger)` is called here https://github.com/mattermost/mattermost-cloud/pull/159/files/1237b244124d3b3fc991f2cb3638d9a90779daf8#diff-4d22c3a7103d8871c0b8cf1a5c67db3cR54 , a cached `logger` from `c` will be used if `c.service` exists, and if not, the passed-in `logger` will be used, when I believe the behavior we desire is for the new logger to always be used when it's passed in. Otherwise, it's impossible to add debugging context at call time, and that's the point of always passing in the logger, I believe.

It's not clear that this is the case in the current version of the code because although `logger` is passed in every time, it is done without adding new debugging context (e.g. `logger.WithFields`) so the functionality appears equivalent.

Reply:
I see what you mean, so should we go back to what was initially? @gabrieljackson Any objection here since you requested this change? https://github.com/mattermost/mattermost-cloud/pull/159#discussion_r389662499

Reply:
No, I think if we want `logger` to be passed in then:
```
if c.service != nil {
c.service.logger = logger
}
```

Right? Otherwise passing in a logger only does something the first time `Session()` is called. If that's the desired behavior @gabrieljackson will have to explain to me why that's desired, I suppose. It doesn't make sense to me why we would ignore the `logger` argument in cases where `c.service` is already initialized.

Reply:
The general idea here is that session may return error but we don't want to panic. Instead, we give a dummy client to the caller which will obviously return an error. This allows the system to keep trying call AWS until a healthy session is constructed. In the case of a failed session, logger is functionally irrelevant anyway. I agree that ignore logger in the case of a cached service but it was the best that I could come up based on the change request previously made. Not sure if this answers your question.

Reply:
It does not. My concern is not about an error case.
Say a user of this method has a `*Client` `c`, and some `logger` in their current scope.
The user calls the method like so:
`c.Session(logger.WithField("foo", "bar")).Something()`
Then, in the same scope, the user calls the method like this:
`c.Session(logger.WithField("bar", "baz")).Something()`
The second call to `Something()` will be called with `c.session` containing field `foo=bar`, and the `c.session` containing fields `bar=baz` will be silently ignored, which is surprising. If this is the desired behavior, I don't understand why that is the case.